### PR TITLE
20 endpoints for subscription

### PR DESCRIPTION
**Pull request by IraMladanovych02**
This PR introduces a new endpoint for creating investments in projects. It includes:

- URL definition following RESTful conventions (POST /api/v1/projects/{project_id}/investment/).
- A new view InvestmentCreateView to handle investment creation requests.
- A serializer InvestmentCreateSerializer with validation logic for investment share.
- Comprehensive tests in InvestmentCreateEndpointTest to verify successful creation, handling of edge cases, and proper error responses.

### CODE REVIEW

File: forum/projects/test/factories.py
Comment (devdsys) [inline forum/projects/test/factories.py:1]: Just to clarify, why 'test' instead of 'tests' for the folder name?

File: forum/projects/views.py [hunk 1]
```diff
@@ -27,3 +35,64 @@ def perform_create(self, serializer):
...
+        data['investor'] = request.user.investor_profile.id
+
+        serializer = InvestmentCreateSerializer(data=request.data)
+        if serializer.is_valid():
+            try:
+                serializer.save(project=project, investor=request.user.investor_profile)
+                return Response(serializer.data, status=status.HTTP_201_CREATED)
+            except ValidationError as e:
+                return Response({"error": str(e)}, status=status.HTTP_400_BAD_REQUEST)
+        return Response(serializer.errors, status=status.HTTP_400_BAD_REQUEST)
```
Comment (mehalyna) [inline forum/projects/views.py:98]: If our logic allows only specific investors to create investments in certain projects, you should ensure that the authenticated investor has the proper permissions to access or invest in the project. For instance:
```
if project.owner == request.user:
    return Response({"error": "You cannot invest in your own project."}, status=status.HTTP_400_BAD_REQUEST)
```

Comment (mehalyna) [inline forum/projects/views.py:98]: Ensure that the share value is within an acceptable range (e.g., between 0 and 100). You can add validation logic either in the serializer or before passing the data to the serializer:
```
if not (0 < data['share'] <= 100):
    return Response({"error": "Share must be between 0 and 100."}, status=status.HTTP_400_BAD_REQUEST)
```

File: forum/projects/test/test_investment_endpoint.py
Comment (mehalyna) [inline forum/projects/test/test_investment_endpoint.py:1]: Ensure the share field does not accept invalid values (e.g., zero or negative amounts).
Add a test case to verify this:
```
def test_create_investment_with_invalid_share(self):
    data = {"share": "0.00"}
    response = self.client.post(self.url, data, format='json')
    self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
    self.assertIn("error", response.data)
```

Comment (mehalyna) [inline forum/projects/test/test_investment_endpoint.py:1]: If the project can only accept investments up to the funding goal, ensure that additional investments beyond the goal are not allowed:
```
def test_create_investment_exceeding_funding_goal(self):
    self.project.funding_goal = Decimal("100.00")
    self.project.save()
    data = {"share": "120.00"}  
    response = self.client.post(self.url, data, format='json')
    self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
    self.assertIn("error", response.data)
```

Comment (mehalyna) [inline forum/projects/test/test_investment_endpoint.py:1]: Add a test to ensure the `is_completed` field is correctly updated when the funding goal is reached:
```
def test_project_completion_after_investment(self):
    data = {"share": "100.00"}
    response = self.client.post(self.url, data, format='json')
    self.assertEqual(response.status_code, status.HTTP_201_CREATED)
    self.project.refresh_from_db()
    self.assertTrue(self.project.is_completed)
```

File: forum/projects/test/test_investment_endpoint.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+        )
+        self.client.force_authenticate(user=self.user)
+        self.url = reverse('projects:project-investment', kwargs={'project_id': self.project.id})
+
+    def test_create_investment_success(self):
+        url = reverse('projects:project-investment', kwargs={"project_id": self.project.id})
+        data = {"share": "50.00"}
+        response = self.client.post(url, data, format='json')
+        self.assertEqual(response.status_code, status.HTTP_201_CREATED)
+        self.assertTrue(Investment.objects.filter(project=self.project, investor=self.investor).exists())
```
Comment (mehalyna) [inline forum/projects/test/test_investment_endpoint.py:46]: Add assertions to validate the content of the response for successful investments

File: forum/projects/test/test_investment_endpoint.py
Comment (mehalyna) [inline forum/projects/test/test_investment_endpoint.py:1]: Test case for invalid `project_id` in the URL:
```
def test_create_investment_invalid_project(self):
    url = reverse('projects:project-investment', kwargs={"project_id": 9999})
    data = {"share": "50.00"}
    response = self.client.post(url, data, format='json')
    self.assertEqual(response.status_code, status.HTTP_404_NOT_FOUND)
```

Comment (mehalyna) [inline forum/projects/test/test_investment_endpoint.py:1]: If only certain roles (e.g., investors) are allowed to invest, add a test to validate permissions:
```
def test_create_investment_as_non_investor(self):
    self.user.role = "startup"  # Set user role to something other than 'investor'
    self.user.save()
    data = {"share": "50.00"}
    response = self.client.post(self.url, data, format='json')
    self.assertEqual(response.status_code, status.HTTP_403_FORBIDDEN)
```

**Review by mehalyna — CHANGES_REQUESTED** [review]

File: forum/projects/test/factories.py
Comment (IraMladanovych02) [reply forum/projects/test/factories.py:1]: done

File: forum/projects/views.py [hunk 3]
```diff
@@ -27,3 +35,64 @@ def perform_create(self, serializer):
...
+
+        data = request.data.copy()
+        data['project'] = project.id
+        data['investor'] = request.user.investor_profile.id
+
+        serializer = InvestmentCreateSerializer(data=request.data)
+        if serializer.is_valid():
+            try:
+                serializer.save(project=project, investor=request.user.investor_profile)
+                return Response(serializer.data, status=status.HTTP_201_CREATED)
```
Comment (SergeKry) [inline forum/projects/views.py:95]: I think we should also check total funding for the project after we created new investment. And if total fundings == funding goal we should mark our project `is_completed = True`. Otherwise projects will never get completed.

File: forum/projects/urls.py [hunk 4]
```diff
@@ -1,9 +1,17 @@
...
 app_name = 'projects'
 
 router = DefaultRouter()
 router.register(r'projects', ProjectViewSet, basename=app_name)
 
-urlpatterns = router.urls
+
+urlpatterns = [
+    path(
+        "api/v1/projects/<int:project_id>/investment/",
```
[the remark this answers is not in the document]
Comment (SergeKry) [reply forum/projects/urls.py:13]: I agree "api/v1/" prefix comes from main urls.py file located in forum directory. In order to make proper url we can either use a router here. But it requires a view to be rewritten (APIView does not support routers out of the box).

Or you can use the approach which worked for auth endpoints. Example can be found in users app.

in `projects/urls.py` we separate router urls and manually described urls.

```
urlpatterns = router.urls

investment_urlpatterns = [
    path(
        "<int:project_id>/investment/",
        InvestmentCreateView.as_view(),
        name="project-investment"
    ),
]
```

in `forum/urls.py`
```

from projects.urls import investment_urlpatterns

api_urlpatterns = [
    path('api/v1/auth/', include((auth_urlpatterns, 'auth'), namespace='auth')),

    #  This line adds you project endpoint with proper prefixes
    path('api/v1/projects/', include((investment_urlpatterns, 'projects'), namespace='projects')),

    *[path(f'api/v1/{app}/', include(urls_file, namespace=app)) for app, urls_file in APP_URLS],
]
```

**Review by SergeKry — CHANGES_REQUESTED** [review]

File: forum/projects/urls.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (IraMladanovych02) [reply forum/projects/urls.py:13]: Done

**Review by mehalyna — APPROVED** [review]

**Review by SergeKry — APPROVED** [review]

**Review by devdsys — APPROVED** [review]
